// see: https://wiki.yandex-team.ru/jandekspoisk/kachestvopoiska/factordev/implement/indexann
package NIndexAnn;
option optimize_for = SPEED;

message TNHopSiteData {
    optional uint32     DwellTime   = 1; // mean dwell time
    optional float      IsFinal     = 2; // mean
    optional uint64     Total       = 3;
    optional uint32     ChainLength = 4; // chain length distribution. Index is chain_length - 1
    optional uint32     Position    = 5; // position distribution
}

message TDummyData {
}

message TBeastQueriesData {
    optional uint32  Position = 1; // Less position - the better.
}

message TFloatStream {
    optional float Value = 1;
}

message TExperimentData {
    /* fields with id 1,3 were uint32 and are unused now */
    optional float   Weight       = 2;
}

message TCorrectedClicksData {
    optional float FRC  = 1;
    optional float UFRC = 2;
}

message TUserQueryUrlData {
    optional float CorrectedCtr = 1;
    optional float SamplePeriodDayFrc = 2;
    optional float YabarVisits  = 3;
    optional float YabarTime = 4;
    optional float CorrectedCtrXFactor = 5;
}

message TClickMachineData {
    reserved 5;
    optional float SplitDt = 1;
    optional float OneClick = 2;
    optional float LongClick = 3;
    optional float SimpleClick = 4;
    optional float LongClickSP = 6;
    optional float RemovedFusClick = 7; // FACTOR-1450
    optional float DoubleFrc = 8;
    optional float OneClickFrcSP = 9;
    optional float QueryDwellTime = 10;
    optional float CorrectedCtrLongPeriod = 11;
    optional float FirstClickDt = 12;
    optional float LongClickMobile = 13;
    optional float FirstLastClickMobile = 14;
    optional float AvgDTWeightedByRankMobile = 15;
    //CorrectedClicks with disabled normalization for vpcg FACTOR-1266, FACTOR-1304
    optional float RawCorrectedClicks = 16;
    optional float VpcgCorrectedClicksSLP = 17;
    //expansions filtering purposes
    optional float RawClicks = 18;
    //source data for expansions
    optional float CorrectedFrcSLP = 19;
    //source data for DSSM
    optional float CorrectedCtrSLP = 20;
    optional float OneClickProb = 21;
    //experimental streams:
    optional float LongClick30SLP = 22;
    optional float LongClick120SLP = 23;
    optional float LongClick120To30RatioSLP = 24;
    optional float LongClick120SqrtSLP = 25;
    optional float LongClick120Sqrt = 26;
}

message TMarketClickMachineData {
    optional float SplitDt = 1;
    optional float OneClick = 2;
    optional float LongClick = 3;
    optional float SimpleClick = 4;
    optional float LongClickMobile = 5;
    optional float FirstLastClickMobile = 6;
    optional float LongClick120Sqrt = 7;
}

message TBannerPhrase {
    optional float Bid = 1;
    optional float HistoryCPM = 2;
    optional float HistoryCPC = 3;
    optional float Ctr = 4;
    optional float Frc = 5;
    optional uint64 TotalShows = 6;
    optional uint64 TotalClicks = 7;
    optional bool IsBroadEnabled = 8;
}

message TBannerOriginalPhrase {
    optional float Enable = 1;
    optional float Bid = 2;
    optional float BidToMaxBid = 3;
}

message TMapsStreams {
    optional float ShowQFRC = 1;
    optional float ClickQFRC = 2;
    optional float OrgAddress = 3;
    optional float OrgName = 4;
    optional float OrgChainName = 5;
    optional float MobileClickQFRC = 6;
    optional float MobileClickSQFRC = 7;
    optional float MobileClickCTR = 8;
    optional float SerpMapsClickQFRC = 9;
    optional float SerpMapsClickSOFRC = 10;
    optional float SerpMapsClickSQFRC = 11;
    optional float LocatedAtOrgName = 12;
    optional float FacultyName = 13;
    optional float HospitalName = 14;
    optional float OrgSummary = 15;
}

message TMarketStreams {
    optional float MarketModelAlias = 1;
    optional TDummyData ImageLinkData = 2;
    optional TDummyData ImageAltData = 3;
    // TShowsTimeDataPB in image ann data
    optional float ImageShowsTime = 4;
    optional float ImageQueryShowsTime = 5;
    // TDwellTimeDataPB in image ann data
    optional float ImageDwellTime = 6;
    optional float ImageQueryDwellTime = 7;
    optional float ImageDocDwellTime = 8;
    // TShowsClicksDataPB in image ann data
    optional float ImageShows = 9;
    optional float ImageClicks = 10;
    optional float ImageQueryClicks = 11;
    optional float MarketDescription = 12;
    optional TDummyData WasOrdered = 13;
    // knn streams
    optional float KNN_5000_9600 = 14;
    optional float KNN_9600_10000 = 15;
    optional float KNN_10000_11000 = 16;
    optional float KNN_11000_20000 = 17;
    optional TDummyData MarketRecipe = 18;
    //
    optional TDummyData BlueDescriptionOfWhiteOffer = 19;
    optional TDummyData BlueTitleOfWhiteOffer = 20;
    optional TDummyData BlueTitleOfWhiteModel = 21;
    optional TDummyData BlueAliasOfWhiteModel = 22;

    optional TDummyData MarketTitle = 23;

    optional TDummyData MarketingDescription = 24;
    optional TDummyData BlueMarketingDescriptionOfModel = 25;
    optional TDummyData BlueMicroDescriptionOfModel = 26;
    optional TDummyData MicroDescriptionOfModel = 27;
    optional TDummyData MskuOfferTitle = 28;
    optional TDummyData MskuOfferSearchText = 29;

    optional TDummyData CPAQuery = 30;
}

message TMarketExpStreams {
    optional float MarketUnoExp = 1;
    optional float MarketDosExp = 2;
    optional float MarketTresExp = 3;
    optional float MarketCuatroExp = 4;
    optional float MarketCincoExp = 5;
    optional float MarketSeisExp = 6;
    optional float MarketSieteExp = 7;
    optional float MarketOchoExp = 8;
    optional float MarketNueveExp = 9;
    optional float MarketDiezExp = 10;
}

message TMarketMetadocStreams {
    optional float BrowserPageRank = 1;
    optional float FirstClickDtXf = 2;
    optional float LongClick = 3;    // ClickMachine
    optional float LongClickSP = 4;    // ClickMachine
    optional float OneClick = 5;    // ClickMachine
    optional float SimpleClick = 6;    // ClickMachine
    optional float SplitDt = 7;    // ClickMachine
    optional float BrowserPageRankSample = 8;
}

message TBurnishCountryData {
    optional float IrrelevantRatio = 1;
    optional float AtLeastRelevantMinusRatio = 2;
    optional float AtLeastRelevantPlusRatio = 3;
}

message TAdvMachineStreams {
    optional float Val1 = 1;
    optional float Val2 = 2;
    optional float Val3 = 3;
    optional float Val4 = 4;
    optional float Val5 = 5;
    optional float Val6 = 6;
    optional float Val7 = 7;
    optional float Val8 = 8;
}

message TAliceMusicStreams {
    optional float TrackTitle       = 1;
    optional float ArtistName       = 2;
    optional float AlbumTitle       = 3;
    optional float TrackArtistNames = 4;
    optional float TrackAlbumTitle  = 5;
    optional float TrackLyrics      = 6;
}

message TMusicStreams {
    optional float FirstClick = 1;
    optional float LastClick = 2;
    optional float LongClick = 3;
    optional float SingleClick = 4;
}

message TRapidClicksStreams {
    optional uint64 Timestamp = 1;
    optional float SearchClicksD30T5 = 2;
}

message TKinopoiskStreams {
    optional float Genre = 1;
    optional float Cast = 2;
    optional float Character = 3;
    optional float Keyword = 4;
    optional float Award = 5;
    optional float Tag = 6;

    optional float Item2ItemSimilarity = 7;
    optional float HandmadeSimilarity = 8;
    optional float PersonCharacter = 9;
}

message TRegionData {
    reserved 33;
    required uint32 Region = 1;

    optional TNHopSiteData             NHop             = 2;
    optional TDummyData                UQ               = 3;
    optional TDummyData                WT               = 4; // FACTOR-312
    optional TBeastQueriesData         BeastQ           = 5; // FACTOR-387
    optional TExperimentData           Experiment       = 6;
    // 7 is not used
    optional TCorrectedClicksData      CorrectedClicks  = 8; // FACTOR-431 & FACTOR-427
    optional TUserQueryUrlData         UserQueryUrl     = 9; // FACTOR-468
    optional TNHopSiteData             NHopVideo        = 10; //FACTOR-534
    optional TFloatStream              BrowserPageRank  = 11; //FACTOR-585
    optional TClickMachineData         ClickMachine     = 12;
    optional TFloatStream              RandomLogDBM35   = 13; //FACTOR-579
    optional TDummyData                DummyData        = 14; // do not change, required for compilation
    optional TFloatStream              PopularSeFrcBrowser = 15; // FACTOR-675
    optional TDummyData                BannerTitle = 16;
    optional TDummyData                BannerText = 17;
    optional TBannerPhrase             BannerPhrase = 18;
    optional TDummyData                BannerMinusWords = 19;
    optional TFloatStream              BannerPhraseXFactor = 20;
    // 21 is not used anymore
    optional TFloatStream              BannerCommQueriesXF = 22;
    optional TFloatStream              VideoQUSM = 23; //VIDEOPOISK-4762
    optional TBannerPhrase             BannerQuery = 24;
    optional TBannerPhrase             BannerRsyaPhrase = 25;
    optional TFloatStream              BrowserPageRankSample  = 26; //FACTOR-799
    optional TFloatStream              OneClickFrcXfSP = 27; // FACTOR-818
    optional TFloatStream              BannerSpyLogTitle = 28; // BSDEV-50854
    optional TFloatStream              LandingPageTitle = 29;
    optional TFloatStream              LandingPageText = 30;
    optional TFloatStream              LandingPageLists = 31;
    optional TFloatStream              NHopSumDwellTime = 32; //FACTOR-811
    optional TFloatStream              VideoMaxClicksPercent = 34;
    optional TFloatStream              VideoPCtrNew = 35;
    optional TFloatStream              VideoPVTMLoki = 36;
    optional TBannerPhrase             BannerGuaranteePhrase = 37;
    optional TBannerPhrase             BannerGuaranteeQuery = 38;
    optional TFloatStream              FirstClickDtXf = 39;
    optional TFloatStream              VideoQTS = 40;
    optional TBannerPhrase             BannerRsyaQuery = 41;
    optional TDummyData                MetaTagSentence = 42;
    optional TFloatStream              VideoRelatedTerm = 43;
    optional TFloatStream              BannerQueryClicksFr = 44;
    optional TFloatStream              BannerRsyaQueryClicksFr = 45;
    optional TFloatStream              BannerQueryCostFr = 46;
    optional TFloatStream              BannerRsyaQueryCostFr = 47;
    optional TMapsStreams              MapsStreams = 48;
    optional TBurnishCountryData       BurnishCountryData = 51; // FACTOR-1293
    optional TBannerOriginalPhrase     BannerOriginalPhrase = 52;
    optional TFloatStream              VideoWideTerm = 53;
    optional TMarketStreams            MarketStreams = 54;

    // Yandex.Market streams BEGIN
    optional TDummyData YaMarketManufacturer = 55;       // param_id=7893318
    optional TDummyData YaMarketCategoryName = 56;
    optional TDummyData YaMarketContents = 57;           // param_id=14885181
    optional TDummyData YaMarketAliases = 58;            // param_id=7351753
    optional TDummyData YaMarketBarcode = 59;            // param_id=14202862
    optional TDummyData YaMarketBookPublisher = 61;      // param_id=15456434
    optional TDummyData YaMarketBookUisbn = 62;          // param_id=15456493
    optional TDummyData YaMarketUrl = 63;                // param_id=7351726
    optional TDummyData YaMarketBookAuthor = 64;         // param_id=15456400
    optional TDummyData YaMarketMobileManufacturer = 65; // param_id=11143646
    optional TFloatStream YaMarketOfferDescription = 66;
    optional TFloatStream YaMarketOfferSalesNotes = 67;
    optional TFloatStream YaMarketOfferTitle = 68;
    optional TFloatStream YaMarketOfferUrl = 69;
    // Yandex.Market streams END
    optional TMarketExpStreams MarketExpStreams = 70;
    optional TMarketClickMachineData MarketClickMachine = 71;
    optional TFloatStream VideohubTerm = 72;
    optional TFloatStream SerNameCRC = 73;
    optional TFloatStream SerNameCRCPlusSeason = 74;
    // Adv Machine
    optional TAdvMachineStreams AdvMachine = 75;

    // Alice Web
    optional TAliceMusicStreams AliceMusicStreams = 76; // BUKI-2932

    optional TFloatStream VideoSpeechToText = 77;

    optional TMusicStreams MusicStreams = 78;

    optional TFloatStream VideoOcr = 79;
    optional TRapidClicksStreams RapidClicksStreams = 80;

    optional TFloatStream VideoClickSim = 81;

    optional TMarketMetadocStreams MarketMetadocStreams = 82;
    optional TNHopSiteData MarketMetadocNHop = 83; 

    optional TKinopoiskStreams KinopoiskStreams = 84;
    optional TFloatStream PrsPosition = 85;

    optional TFloatStream VideoHashtags = 86;
    optional TFloatStream VideoTranslatedTitle = 87;
    optional TFloatStream VideoTranslatedText = 88;
}

message TAnnotationRec {
    optional string         Text   = 1; // in indexhanndata, contains *wide* text
    repeated TRegionData    Data   = 2; // sorted by region
    optional uint32         TextLanguage = 5; // recognized language of Text
}

message TIndexAnnSiteData {
    repeated TAnnotationRec Recs = 1;
    optional bool IsOnlyFastData = 2; // from fast userdata, used for correct aggregation of long and fast factors
}

message TIndexAnnHostData {
    required string            Host = 1;
    optional TIndexAnnSiteData Data = 2;
}
